plugin-server/src/utils/retries.ts

```diff
@@ -0,0 +1,112 @@
+import { RetryError } from '@posthog/plugin-scaffold'
+
+import { runInTransaction } from '../sentry'
+import { Hub } from '../types'
+
+export function getNextRetryMs(baseMs: number, multiplier: number, attempt: number): number {
+    if (attempt < 1) {
+        throw new Error('Attempts are indexed starting with 1')
+    }
+    return baseMs * multiplier ** (attempt - 1)
+}
+
+export interface RetriableFunctionDefinition {
+    payload: Record<string, any>
+    tryFn: () => void | Promise<void>
+    catchFn?: (error: Error | RetryError) => void | Promise<void>
+    finallyFn?: (attempts: number) => void | Promise<void>
+}
+
+export interface RetryParams {
+    maxAttempts: number
+    retryBaseMs: number
+    retryMultiplier: number
+}
+
+export interface MetricsDefinition {
+    metricPrefix: string
+    metricName: string
+    metricTags?: Record<string, string>
+}
+
+export type RetriableFunctionPayload = RetriableFunctionDefinition &
+    Partial<RetryParams> &
+    MetricsDefinition & { hub: Hub }
+
+function iterateRetryLoop(retriableFunctionPayload: RetriableFunctionPayload, attempt = 1): Promise<void> {
+    const {
+        metricPrefix,
+        metricName,
+        metricTags = {},
+        hub,
+        payload,
+        tryFn,
+        catchFn,
+        finallyFn,
+        maxAttempts = 5,
+        retryBaseMs = 5000,
+        retryMultiplier = 2,
+    } = retriableFunctionPayload
+    return runInTransaction(
+        {
+            name: 'retryLoop',
+            op: metricName,
+            description: metricTags.plugin || '?',
+            data: {
+                metricName,
+                payload,
```

Comment:
Does nesting objects like this work nicely in sentry?

Reply:
Unsure - again, this was here before, just it was named "event"